syntax = "proto3";

package postservice;

service PostService {
    rpc CreatePost(CreatePostRequest) returns (PostResponse);
    rpc DeletePost(DeletePostRequest) returns (EmptyResponse);
    rpc UpdatePost(UpdatePostRequest) returns (PostResponse);
    rpc GetPost(GetPostRequest) returns (PostResponse);
    rpc ListPosts(ListPostsRequest) returns (ListPostsResponse);
}

message Post {
    string id = 1;
    string title = 2;
    string description = 3;
    string creator_id = 4;
    string created_at = 5;
    string updated_at = 6;
    bool is_private = 7;
    repeated string tags = 8;
}

message CreatePostRequest {
    string title = 1;
    string description = 2;
    string creator_id = 3;
    bool is_private = 4;
    repeated string tags = 5;
}

message DeletePostRequest {
    string post_id = 1;
    string requester_id = 2;
}

message UpdatePostRequest {
    string post_id = 1;
    string title = 2;
    string description = 3;
    bool is_private = 4;
    repeated string tags = 5;
    string requester_id = 6;
}

message GetPostRequest {
    string post_id = 1;
    string requester_id = 2;
}

message ListPostsRequest {
    int32 page = 1;
    int32 page_size = 2;
    string requester_id = 3;
}

message PostResponse {
    Post post = 1;
}

message ListPostsResponse {
    repeated Post posts = 1;
    int32 total = 2;
}

message EmptyResponse {}
